readerwriterlock/rwlock_async.py

```diff
@@ -0,0 +1,816 @@
+			"""Acquire a lock."""
+			p_timeout: Optional[float] = None if (blocking and timeout < 0) else (timeout if blocking else 0)
+			c_deadline: Optional[float] = None if p_timeout is None else (self.c_rw_lock.c_time_source() + p_timeout)
+			try:
+				await asyncio.wait_for(self.c_rw_lock.c_lock_read_count.acquire(), timeout=(None if c_deadline is None else max(sys.float_info.min, c_deadline - self.c_rw_lock.c_time_source())))
+			except asyncio.TimeoutError:
+				return False
+			await self.c_rw_lock.v_read_count.inc()
+			if 1 == int(self.c_rw_lock.v_read_count):
+				try:
+					await asyncio.wait_for(self.c_rw_lock.c_resource.acquire(), timeout=(None if c_deadline is None else max(sys.float_info.min, c_deadline - self.c_rw_lock.c_time_source())))
+				except asyncio.TimeoutError:
+					await self.c_rw_lock.v_read_count.dec()
+					self.c_rw_lock.c_lock_read_count.release()
+					return False
+			self.c_rw_lock.c_lock_read_count.release()
+			self.v_locked = True
+			return True
+
+		async def release(self) -> None:
+			"""Release the lock."""
+			if not self.v_locked: raise RELEASE_ERR_CLS(RELEASE_ERR_MSG)
+			self.v_locked = False
+			await self.c_rw_lock.c_lock_read_count.acquire()
+			await self.c_rw_lock.v_read_count.dec()
+			if 0 == int(self.c_rw_lock.v_read_count):
+				self.c_rw_lock.c_resource.release()
+			self.c_rw_lock.c_lock_read_count.release()
+
+		def locked(self) -> bool:
+			"""Answer to 'is it currently locked?'."""
+			return self.v_locked
+
+	class _aWriter(LockableD):
+		def __init__(self, p_RWLock: "RWLockReadD") -> None:
+			self.c_rw_lock = p_RWLock
+			self.v_locked: bool = False
+			super().__init__()
+
+		async def acquire(self, blocking: bool = True, timeout: float = -1) -> bool:
+			"""Acquire a lock."""
+			p_timeout: Optional[float] = None if (blocking and timeout < 0) else (timeout if blocking else 0)
+			c_deadline: Optional[float] = None if p_timeout is None else (self.c_rw_lock.c_time_source() + p_timeout)
+			try:
+				await asyncio.wait_for(self.c_rw_lock.c_resource.acquire(), timeout=(None if c_deadline is None else max(sys.float_info.min, c_deadline - self.c_rw_lock.c_time_source())))
+				locked: bool = True
+			except asyncio.TimeoutError:
+				locked: bool = False
+
+			self.v_locked = locked
+			return locked
+
+		async def downgrade(self) -> Lockable:
+			"""Downgrade."""
+			if not self.v_locked: raise RELEASE_ERR_CLS(RELEASE_ERR_MSG)
+
+			result = self.c_rw_lock.gen_rlock()
+
+			wait_blocking = asyncio.Event()
+
+			async def lock_result() -> None:
+				wait_blocking.set()
+				await result.acquire()  # This is a blocking action
+				wait_blocking.set()
+
+			run_task(lock_result())
+
+			await wait_blocking.wait()  # Wait for the thread to be almost in its blocking state.
+			wait_blocking.clear()  # TODO is this needed?
+
+			await asyncio.sleep(sys.float_info.min * 123)  # Heuristic sleep delay to leave some extra time for the thread to block.
```

Comment:
Yeah, I wrestled with this function for a while before I came up with the current implementation. Here is my idea of the sequence of events:

1. Line 525: Schedule `lock_result()`
2. Line 532: Give up control until until `lock_result()` has been scheduled and started executing (in case the implementation doesn't start executing right away on line 530)
3. Line 526: Set the event
4. Line 527: Attempt to acquire the lock
5. Line 533: Clear the event (I forgot to check if that was required or not)
6. Line 535: Give up execution again to try and ensure that `lock_result()` was able to get through all the `await`'s in `RWLockReadD._aReader.acquire()`
7. Line 537: Release the writer lock
8. Line 539: wait for the event to trigger
9. Line 528: Set the event
10. Line 540: return the result

Honestly, it might be smarter to do multiple sleeps. That way we give up our time slice multiple times to give the best chance for `lock_result()` to get to the `c_resource` lock.

I honestly debated for a while if I should add an optional parameter to ` `RWLockReadD._aReader.acquire()` which would let an event be passed in. `acquire()` would then set the event right before it blocked on the `c_resource` lock, giving the best chances of unlocking at the right time.

What do you think?

Reply:
I think that a perfect solution does not exists **in theory**.
* It does not exists with non-async implementation
* And it does not exists with async implementation

But **in practice**, in non-async, unit test have been running since a while and it never managed to enter this corner case and this package has been downloaded million of times and nobody ever reported an issue about writer sneaking in when downgrading.

I think that the combination of (multi sleep) + (the way the python interpreter scheduler is internally implemented) = Multi sleep is a pretty good solution

Your idea about the optional parameter + signal... etc is a good idea which would certainly improve the situation since it guarantee that some of the await are done and it would be preventing let say 50% of the time this corner case would occurs, which is better but still a non-perfect solution… and 50% of it never happened yet = I am not sure it is worth the overhead.

**Conclusion:**
If you want to implement the optional parameter + signal, it is a good idea, go for it, else put multi-sleep.

I'll let you decide… no matter which solution you chose, I'll merge your pull request.

Reply:
Hello @linted,

I looked for information about python internal handling of thread and sleep method, etc…, this is what I found:

* It isn't possible in python to set higher/lower priority to a thread
* Releasing the python GIL (*may* allow another thread to do stuff
* Sleep call with a parameter of 0 aren't ignored, they do release the GIL
* Sleep call with a parameter of 0 has a high chance of choosing the *same* thread again
* Sleep call with a non-zero parameter release the GIL **and** do a system call (system call mean that the OS is notified and could give cycle time to another process)


With those details about the behavior of the python scheduler, using a bunch of sleep call with a non-zero value really seems to be the best solution (this solution works and is easy to implement, easy to maintain).

Reply:
I think I agree that the better solution will probably be to call sleep a couple of time, just to make sure that everything is good.

While python's asyncio could be implemented with threading, i don't think it is required to do so, and is implementation specific. Technically everything could be implemented in a single thread, albeit very inefficiently in my opinion. I think the best way to think about it is to assume that every line containing an `await` is effectively handing over your time slice.

To illustrate this, I wrote a simple little program that will run a single itteration of the event loop and print out inbetween each itteration. This works really well if you place prints surrounding each await call.

```python
#!/usr/bin/env python3
import asyncio
from readerwriterlock import rwlock_async
eloop = asyncio.get_event_loop()
locker = rwlock_async.RWLockReadD()


async def main():
  wlock = locker.gen_wlock()
  await wlock.acquire()
  asyncio.create_task(wlock.downgrade())

eloop.create_task(main())
for i in range(20):
  print("Doing round {}".format(i))
  eloop.stop()
  eloop.run_forever()
```

This results in the following output:
```
Doing round 0
Doing round 1
Line 534: before run_task(lock_result())
Line 536: after run_task(lock_result())
Line 638: before wait_blocking.wait
Doing round 2
Line 526: before wait_blocking.set
Line 528: before result.acquire
Doing round 3
Line 541: after wait_blocking.clear
Line 543: before sleep
Line 461: before c_lock_read_count.acquire
Line 463: after c_lock_read_count.acquire
Line 466: before v_read_count.increment
Line 468: after v_read_count.increment
Line 471: before c_resource.acquire
Doing round 4
Doing round 5
Line 546: before self.release
Line 548: after self.release
Doing round 6
Line 473: after c_resource.acquire
Doing round 7
Line 530: after result.acquire
Line 532: done lock_result
Doing round 8
Line 550: after wait_blocking.wait
Doing round 9
Doing round 10
Doing round 11
Doing round 12
Doing round 13
Doing round 14
Doing round 15
Doing round 16
Doing round 17
Doing round 18
Doing round 19
```

So if we assume that there is contention for the `c_lock_read_count` lock or the lock in `v_read_count`, they could get delayed to rounds sometime after round 5. Having multiple calls to sleep will, to my knowledge, push back the execute by at least 1 round per sleep call.